File: python_sdk/src/reality_capture/specifications/volume.py
from pydantic import BaseModel, Field
from typing import Optional
from enum import Enum


class VolumeInputs(BaseModel):
    model_3d: str = Field(alias="model3d", description="Reality data id of a point cloud.")
    footprint: str = Field(alias="footprint", pattern=r"^bkt:.+",
                           description="Path in the bucket to footprint for volume computation.")


class VolumeOutputs(BaseModel):
    volume: Optional[str] = Field(None, description="Path in the bucket to volume information.", pattern=r"^bkt:.+")


class VolumeOutputsCreate(Enum):
    VOLUME = "volume"


class VolumeSpecificationsCreate(BaseModel):
    inputs: VolumeInputs = Field(description="Inputs")
    outputs: list[VolumeOutputsCreate] = Field(description="Outputs")


class VolumeSpecifications(BaseModel):
    inputs: VolumeInputs = Field(description="Inputs")
    outputs: VolumeOutputs = Field(description="Outputs")

File: typescript/packages/reality-capture/src/specifications/volume.ts
import { z } from "zod";

export const VolumeInputsSchema = z.object({
  model3d: z.string().describe("Reality data id of a point cloud."),
  regionOfInterest: z.string().regex(/^bkt:.+/).describe("Path in the bucket to region of interest for volume computation."),
});
export type VolumeInputs = z.infer<typeof VolumeInputsSchema>;

export const VolumeOutputsSchema = z.object({
  volume: z.string().regex(/^bkt:.+/).describe("Path in the bucket to volume information.").optional(),
});
export type VolumeOutputs = z.infer<typeof VolumeOutputsSchema>;

export enum VolumeOutputsCreate {
  VOLUME = "volume",
}

export const VolumeSpecificationsCreateSchema = z.object({
  inputs: VolumeInputsSchema.describe("Inputs"),
  outputs: z.array(z.nativeEnum(VolumeOutputsCreate)).describe("Outputs"),
});
export type VolumeSpecificationsCreate = z.infer<typeof VolumeSpecificationsCreateSchema>;

export const VolumeSpecificationsSchema = z.object({
  inputs: VolumeInputsSchema.describe("Inputs"),
  outputs: VolumeOutputsSchema.describe("Outputs"),
});
export type VolumeSpecifications = z.infer<typeof VolumeSpecificationsSchema>;
